### PR TITLE
Setup basic API

### PR DESCRIPTION
**Pull request by ellie**
Needs routes to hook into pgqueue obv, but

1. Pull out common dependencies, either currently common or likely future common
2. Add http metrics
3. Start up axum server
4. Add config

### CODE REVIEW

File: hook-producer/src/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            Matcher::Full("http_requests_duration_seconds".to_string()),
+            EXPONENTIAL_SECONDS,
+        )
+        .unwrap()
+        .install_recorder()
+        .unwrap()
+}
+
+/// Middleware to record some common HTTP metrics
+/// Someday tower-http might provide a metrics middleware: https://github.com/tower-rs/tower-http/issues/57
```
Comment (bretthoerner) [inline hook-producer/src/metrics.rs:24]: Nice, thanks for the link, subscribed.

File: hook-producer/src/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        matched_path.as_str().to_owned()
+    } else {
+        req.uri().path().to_owned()
+    };
+
+    let method = req.method().clone();
+
+    // Run the rest of the request handling first, so we can measure it and get response
+    // codes.
+    let response = next.run(req).await;
```
Comment (bretthoerner) [inline hook-producer/src/metrics.rs:38]: Huh, the way they do middleware is cool.

**Review by bretthoerner — APPROVED** [review]

**Comment by bretthoerner** [comment]
Hm, I didn't expect the `Unresolved conversations` checks. That seems kind of annoying.

File: hook-producer/src/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    let latency = start.elapsed().as_secs_f64();
+    let status = response.status().as_u16().to_string();
```
Comment (tomasfarias) [inline hook-producer/src/metrics.rs:41]: I really wonder why `as_str` goes through the trouble of [figuring an offset](https://docs.rs/http/latest/src/http/status.rs.html#146) to get the string representation of the status code from a [hardcoded string of codes](https://docs.rs/http/latest/src/http/status.rs.html#543).

Comment (tomasfarias) [reply hook-producer/src/metrics.rs:41]: Maybe it's so it doesn't need to allocate an additional `String`?

Comment (tomasfarias) [reply hook-producer/src/metrics.rs:41]: Anyways, this is perfectly fine.

File: hook-producer/src/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+use std::time::Instant;
+
+use axum::{
+    body::Body, extract::MatchedPath, http::Request, middleware::Next, response::IntoResponse,
+};
+use metrics_exporter_prometheus::{Matcher, PrometheusBuilder, PrometheusHandle};
+
+pub fn setup_metrics_recorder() -> PrometheusHandle {
+    const EXPONENTIAL_SECONDS: &[f64] = &[
+        0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
```
Comment (tomasfarias) [inline hook-producer/src/metrics.rs:10]: question: Where does this "x2/x2.5" series come from?

File: hook-producer/src/metrics.rs (hunk 1 above)
Comment (ellie) [reply hook-producer/src/metrics.rs:24]: no problem

**Comment by ellie** [comment]
> Hm, I didn't expect the Unresolved conversations checks. That seems kind of annoying.

Oof yeah, I'll disable it

File: hook-producer/src/metrics.rs (hunk 4 above)
Comment (ellie) [reply hook-producer/src/metrics.rs:10]: It's a series I've copy-pasted from my [snippet](https://ellie.wtf/notes/exporting-prometheus-metrics-with-axum) for a while now, I think originating from the axum example docs. Generally seems to work quite well, no issues if it needs changing.

Comment (tomasfarias) [reply hook-producer/src/metrics.rs:10]: Cool! Thank you! I was just curious where it came from. I think it works fine!

**Pull request merged**